ffi/memorymanager.cpp

```diff
@@ -0,0 +1,355 @@
+// SPDX-License-Identifier: Apache-2.0 WITH LLVM-exception
+//
+//===----------------------------------------------------------------------===//
+//
+// This file implements the section-based memory manager used by the MCJIT
+// execution engine and RuntimeDyld
+//
+//===----------------------------------------------------------------------===//
+
+#include "memorymanager.h"
+#include "llvm/Support/MathExtras.h"
+#include "llvm/Support/Process.h"
+
+#define DEBUG_TYPE "llvmlite-memory-manager"
+
+namespace llvm {
+
+uint8_t *LlvmliteMemoryManager::allocateDataSection(uintptr_t Size,
+                                                    unsigned Alignment,
+                                                    unsigned SectionID,
+                                                    StringRef SectionName,
+                                                    bool IsReadOnly) {
+    if (IsReadOnly)
+        return allocateSection(LlvmliteMemoryManager::AllocationPurpose::ROData,
+                               Size, Alignment);
+    return allocateSection(LlvmliteMemoryManager::AllocationPurpose::RWData,
+                           Size, Alignment);
+}
+
+uint8_t *LlvmliteMemoryManager::allocateCodeSection(uintptr_t Size,
+                                                    unsigned Alignment,
+                                                    unsigned SectionID,
+                                                    StringRef SectionName) {
+    return allocateSection(LlvmliteMemoryManager::AllocationPurpose::Code, Size,
+                           Alignment);
+}
+
+uint8_t *LlvmliteMemoryManager::allocateSection(
+    LlvmliteMemoryManager::AllocationPurpose Purpose, uintptr_t Size,
+    unsigned Alignment) {
+    if (!Alignment)
+        Alignment = 16;
+
+    LLVM_DEBUG(dbgs() << "Requested size / alignment: "
+                      << format_hex(Size, 2, true) << " / " << Alignment
+                      << "\n");
+
+    assert(!(Alignment & (Alignment - 1)) &&
+           "Alignment must be a power of two.");
+
+    uintptr_t RequiredSize =
+        Alignment * ((Size + Alignment - 1) / Alignment + 1);
+    uintptr_t Addr = 0;
+
+    LLVM_DEBUG(dbgs() << "Allocating " << format_hex(RequiredSize, 2, true)
+                      << " bytes for ");
+
+    MemoryGroup &MemGroup = [&]() -> MemoryGroup & {
+        switch (Purpose) {
+        case AllocationPurpose::Code:
+            LLVM_DEBUG(dbgs() << "CodeMem at ");
+            return CodeMem;
+        case AllocationPurpose::ROData:
+            LLVM_DEBUG(dbgs() << "RODataMem at ");
+            return RODataMem;
+        case AllocationPurpose::RWData:
+            LLVM_DEBUG(dbgs() << "RWDataMem at ");
+            return RWDataMem;
+        }
+        llvm_unreachable("Unknown LlvmliteMemoryManager::AllocationPurpose");
+    }();
+
+    // Look in the list of free memory regions and use a block there if one
+    // is available.
+    for (FreeMemBlock &FreeMB : MemGroup.FreeMem) {
```

Comment:
About 

>  I don't understand how the memory allocation / mapping really works, in particular what pending memory is and pending prefix indices are 

My understanding is that pending memory is memory that has allocated but not yet "finalised".
I am not sure how importing this prefix index is. It looks like a bit of bookkeeping to keep an index of the next free block. 

Also my impression is that this whole loop is skipped because of clearance on lines 137 - 139.